### PR TITLE
Add developer documentation forbidding default parameters in detail APIs

### LINKED ISSUE #12944: [FEA] Remove default parameters from detail APIs
**Is your feature request related to a problem? Please describe.**
In numerous instances we have found errors in implementations failing to pass a stream or memory resource all the way down the call stack, resulting in, for instance, memory being allocated using the wrong resource. #9854 discusses the benefits of requiring callers of detail APIs to always provide explicit streams, which obviates the possibility for this sort of oversight. We should make the same change for memory resources to ensure that they are always properly propagated down.

**Describe the solution you'd like**
We should systematically remove all default mr parameters from detail APIs. Since stream and mr are always the last two parameters for every function, this will require removing all other default parameters as well. While this task sounds extensive, in practice the number of non-mr default parameters remaining in detail APIs is quite small (completing #9854 got us most of the way there on other parameters).

See discussion on #9584 and #11967 for more context.

### PR DESCRIPTION
**Pull request by vyasr**
## Description

Resolves #12944. Also reflows the text to ensure consistent text width.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by vyasr** [comment]
Technically PRs #12964-#12970 are also required to close the associated issue, but I am assuming that these will be merged within a reasonable amount of time of each other.

**Review by davidwendt — APPROVED** [review]

**Review by PointKernel — APPROVED** [review]

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md [hunk 1]
```diff
@@ -617,6 +609,30 @@ rmm::mr::device_memory_resource * mr = new my_custom_resource{...};
...
+Default memory resource parameters make it easy for developers to accidentally allocate memory using the incorrect resource.
+Avoiding default memory resources forces developers to consider each memory allocation carefully.
+
+While streams are not currently exposed in libcudf's API, we plan to do so eventually.
+As a result, the same arguments for memory resources also apply to streams.
+Public APIs will default to using `cudf::get_default_stream()`.
+However, including the same default in detail APIs opens the door for developers to forget to pass in a user-provided stream if one is passed to a public API.
+Forcing every API to explicitly pass the stream is intended to prevent such mistakes.
+
+The memory resources -- and eventually, the stream -- are the final parameters for essentially all public APIs.
```
Comment (harrism) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:624]: [This is just pet-peevery] Github Markdown doesn't automatically render emdashes (and these should be mdash `---` not endash `--`) unless you use the HTML code `&mdash`.  Luckily this works just fine with parentheses.

```suggestion
The memory resource (and eventually the stream) is the final parameter for essentially all public APIs.
```

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md [hunk 2]
```diff
@@ -617,6 +609,30 @@ rmm::mr::device_memory_resource * mr = new my_custom_resource{...};
...
+For API consistency, the same is true throughout libcudf's internals.
+Therefore, a consequence of not allowing default streams or mrs is that no parameters in detail APIs may have defaults.
```
Comment (harrism) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:626]:
```suggestion
Therefore, a consequence of not allowing default streams or MRs is that no parameters in detail APIs may have defaults.
```

**Review by harrism — APPROVED** [review]

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (vyasr) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:626]: Seems like the consensus line width is 100 columns, with a few violations. I reflowed everything that was off.

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md (hunk 1 above)
Comment (vyasr) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:624]: I wasn't sure about dash rendering either! I left off the third hyphen accidentally, but I agree that a parenthetical is preferable here anyway.

**Review by bdice — APPROVED** [review]

**Comment by bdice** [comment]
This PR is too far behind `branch-23.04` to pass the Recently Updated check. I'll update it so that CI can merge when it passes.

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md]: Yeah, 100, correct. Thanks for reflowing.

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md [hunk 3]
```diff
@@ -617,6 +633,32 @@ rmm::mr::device_memory_resource * mr = new my_custom_resource{...};
...
+The memory resources (and eventually, the stream) are the final parameters for essentially all
+public APIs. For API consistency, the same is true throughout libcudf's internals. Therefore, a
+consequence of not allowing default streams or MRs is that no parameters in detail APIs may have
+defaults.
+
+## NVTX Ranges
+
+In order to aid in performance optimization and debugging, all compute intensive libcudf functions
+should have a corresponding NVTX range. libcudf has a convenience macro `CUDF_FUNC_RANGE()` that
+will automatically annotates the lifetime of the enclosing function and uses the function's name as
```
Comment (harrism) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:658]:
```suggestion
automatically annotates the lifetime of the enclosing function and uses the function's name as
```

**Pull request merged**